### PR TITLE
PP-5360 get one transaction

### PR DESCRIPTION
**Pull request by jankowiakmaria**
depends on https://github.com/alphagov/pay-publicapi/pull/580

### CODE REVIEW

File: src/main/java/uk/gov/pay/ledger/transaction/model/PaymentFactory.java [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+
+    @Inject
+    public PaymentFactory(ObjectMapper objectMapper) {
+        this.objectMapper = objectMapper;
+    }
+
+    public Payment fromTransactionEntity(TransactionEntity entity) {
+        try {
+            JsonNode transactionDetails = objectMapper.readTree(Optional.ofNullable(entity.getTransactionDetails()).orElse("{}"));
+            Address billingAddress = Address.from(
```
Comment (sfount) [inline src/main/java/uk/gov/pay/ledger/transaction/model/PaymentFactory.java]: This could be a constructor? It seems to be directly passing in params to make an Address object which doesn't seem like the right fit for a `from` pattern

File: src/main/java/uk/gov/pay/ledger/transaction/model/PaymentFactory.java [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+                    safeGetAsString(transactionDetails, "address_line1"),
+                    safeGetAsString(transactionDetails, "address_line2"),
+                    safeGetAsString(transactionDetails, "address_postcode"),
+                    safeGetAsString(transactionDetails, "address_city"),
+                    safeGetAsString(transactionDetails, "address_county"),
+                    safeGetAsString(transactionDetails, "address_country")
+            );
+
+            CardDetails cardDetails = CardDetails.from(entity.getCardholderName(), billingAddress, entity.getCardBrand(),
```
Comment (sfount) [inline src/main/java/uk/gov/pay/ledger/transaction/model/PaymentFactory.java]: Again not sure about `from` pattern, I might be wrong on this!

File: src/main/java/uk/gov/pay/ledger/transaction/model/PaymentFactory.java [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+                    TransactionState.from(entity.getState()), safeGetAsString(transactionDetails, "language"),
+                    entity.getExternalId(), safeGetAsString(transactionDetails, "return_url"), entity.getEmail(),
+                    safeGetAsString(transactionDetails, "payment_provider"), entity.getCreatedDate(),
+                    cardDetails, Boolean.valueOf(safeGetAsString(transactionDetails, "delayed_capture")),
+                    metadata, entity.getEventCount(), safeGetAsString(transactionDetails, "gateway_transaction_id"),
+                    safeGetAsLong(transactionDetails, "corporate_surcharge"), safeGetAsLong(transactionDetails, "fee"),
+                    entity.getNetAmount(), refundSummary, entity.getTotalAmount(), settlementSummary);
+        } catch (IOException e) {
+            LOGGER.error(format("Error during the parsing transaction entity data {}", entity.getExternalId()), e);
+            throw new RuntimeException(e);
```
Comment (sfount) [inline src/main/java/uk/gov/pay/ledger/transaction/model/PaymentFactory.java]: Is there something that catches and deals with this exception? It might be outside of this PR but if there isn't it might not need to be rethrown

File: src/main/java/uk/gov/pay/ledger/transaction/search/model/SettlementSummary.java [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+public class SettlementSummary {
+    private ZonedDateTime captureSubmitTime, capturedTime;
+
+    public SettlementSummary(ZonedDateTime settlementSubmittedTime, ZonedDateTime settledTime) {
+        captureSubmitTime = settlementSubmittedTime;
+        capturedTime = settledTime;
+    }
+
+    public static SettlementSummary ofValue(ZonedDateTime settlementSubmittedTime, ZonedDateTime settledTime) {
+        return new SettlementSummary(settlementSubmittedTime, settledTime);
```
Comment (sfount) [inline src/main/java/uk/gov/pay/ledger/transaction/search/model/SettlementSummary.java]: As this constructor is `public` could the code calling `ofValue` not directly instantiate the object?

File: src/main/java/uk/gov/pay/ledger/transaction/search/model/SettlementSummary.java [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+import java.time.format.DateTimeFormatter;
+import java.util.Objects;
+
+import static uk.gov.pay.commons.model.ApiResponseDateTimeFormatter.ISO_INSTANT_MILLISECOND_PRECISION;
+
+public class SettlementSummary {
+    private ZonedDateTime captureSubmitTime, capturedTime;
+
+    public SettlementSummary(ZonedDateTime settlementSubmittedTime, ZonedDateTime settledTime) {
+        captureSubmitTime = settlementSubmittedTime;
```
Comment (sfount) [inline src/main/java/uk/gov/pay/ledger/transaction/search/model/SettlementSummary.java]: Is there a reason these have different names, are they equivalent? It might be clearer to have standard naming and then explicitly rename in the `@JsonProperty` or something

File: src/main/java/uk/gov/pay/ledger/transaction/search/model/SettlementSummary.java [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+        capturedTime = settledTime;
+    }
+
+    public static SettlementSummary ofValue(ZonedDateTime settlementSubmittedTime, ZonedDateTime settledTime) {
+        return new SettlementSummary(settlementSubmittedTime, settledTime);
+    }
+
+    @JsonProperty("capture_submit_time")
+    public String getCaptureSubmitTime() {
+        return (captureSubmitTime != null) ? ISO_INSTANT_MILLISECOND_PRECISION.format(captureSubmitTime) : null;
```
Comment (sfount) [inline src/main/java/uk/gov/pay/ledger/transaction/search/model/SettlementSummary.java]: Could this be an `Optional<String>`

File: src/main/java/uk/gov/pay/ledger/transaction/entity/TransactionEntity.java [hunk 7]
```diff
@@ -37,14 +37,26 @@
...
+
     public TransactionEntity() {}
 
     public TransactionEntity(Long id, String gatewayAccountId, String externalId,
                              Long amount, String reference, String description,
                              String state, String email, String cardholderName,
                              String externalMetadata, ZonedDateTime createdDate,
                              String transactionDetails, Integer eventCount, String cardBrand,
-                             String lastDigitsCardNumber, String firstDigitsCardNumber) {
+                             String lastDigitsCardNumber, String firstDigitsCardNumber,
```
[the remark this answers is not in the document]
Comment (jankowiakmaria) [reply src/main/java/uk/gov/pay/ledger/transaction/entity/TransactionEntity.java]: Agreed. Is it fine to have it in separate PR? (didn't want to extend the size of this one even more)

File: src/main/java/uk/gov/pay/ledger/transaction/model/PaymentFactory.java [hunk 8]
```diff
@@ -0,0 +1,84 @@
...
+
+    private static Logger LOGGER = LoggerFactory.getLogger(PaymentFactory.class);
+    private ObjectMapper objectMapper;
+
+    @Inject
+    public PaymentFactory(ObjectMapper objectMapper) {
+        this.objectMapper = objectMapper;
+    }
+
+    public Payment fromTransactionEntity(TransactionEntity entity) {
```
[the remark this answers is not in the document]
Comment (jankowiakmaria) [reply src/main/java/uk/gov/pay/ledger/transaction/model/PaymentFactory.java]: fair enough

File: src/main/java/uk/gov/pay/ledger/transaction/model/PaymentFactory.java [hunk 9]
```diff
@@ -0,0 +1,84 @@
...
+            return new Payment(entity.getGatewayAccountId(), entity.getAmount(), entity.getReference(), entity.getDescription(),
+                    TransactionState.from(entity.getState()), safeGetAsString(transactionDetails, "language"),
+                    entity.getExternalId(), safeGetAsString(transactionDetails, "return_url"), entity.getEmail(),
+                    safeGetAsString(transactionDetails, "payment_provider"), entity.getCreatedDate(),
+                    cardDetails, Boolean.valueOf(safeGetAsString(transactionDetails, "delayed_capture")),
+                    metadata, entity.getEventCount(), safeGetAsString(transactionDetails, "gateway_transaction_id"),
+                    safeGetAsLong(transactionDetails, "corporate_surcharge"), safeGetAsLong(transactionDetails, "fee"),
+                    entity.getNetAmount(), refundSummary, entity.getTotalAmount(), settlementSummary);
+        } catch (IOException e) {
+            LOGGER.error(format("Error during the parsing transaction entity data {}", entity.getExternalId()), e);
```
[the remark this answers is not in the document]
Comment (jankowiakmaria) [reply src/main/java/uk/gov/pay/ledger/transaction/model/PaymentFactory.java]: ok

File: src/main/java/uk/gov/pay/ledger/transaction/model/PaymentFactory.java [hunk 10]
```diff
@@ -0,0 +1,84 @@
...
+            CardDetails cardDetails = CardDetails.from(entity.getCardholderName(), billingAddress, entity.getCardBrand(),
+                    entity.getLastDigitsCardNumber(), entity.getFirstDigitsCardNumber(),
+                    safeGetAsString(transactionDetails, "card_expiry_date"));
+
+            Map<String, Object> metadata = null;
+            if(entity.getExternalMetadata() != null) {
+                metadata = objectMapper.readValue(entity.getExternalMetadata(), new TypeReference<Map<String, Object>>() {});
+            }
+
+            RefundSummary refundSummary = RefundSummary.ofValue(entity.getRefundStatus(), entity.getRefundAmountAvailable(), entity.getRefundAmountSubmitted());
```
[the remark this answers is not in the document]
Comment (jankowiakmaria) [reply src/main/java/uk/gov/pay/ledger/transaction/model/PaymentFactory.java]: Can do. Not sure which way is cleaner.

File: src/main/java/uk/gov/pay/ledger/transaction/model/PaymentFactory.java (hunk 1 above)
Comment (jankowiakmaria) [reply src/main/java/uk/gov/pay/ledger/transaction/model/PaymentFactory.java]: It can't be constructor. The from method might return `null` and constructors can't return `null`.
https://github.com/alphagov/pay-ledger/blob/master/src/main/java/uk/gov/pay/ledger/transaction/model/Address.java#L30-L41

File: src/main/java/uk/gov/pay/ledger/transaction/model/PaymentFactory.java (hunk 2 above)
Comment (jankowiakmaria) [reply src/main/java/uk/gov/pay/ledger/transaction/model/PaymentFactory.java]: Again it can't.
https://github.com/alphagov/pay-ledger/blob/master/src/main/java/uk/gov/pay/ledger/transaction/model/CardDetails.java#L30-L42

Comment (sfount) [reply src/main/java/uk/gov/pay/ledger/transaction/model/PaymentFactory.java]: Understood! Maybe having this as optional throughout the code would make this clear

File: src/main/java/uk/gov/pay/ledger/transaction/dao/mapper/TransactionMapper.java [hunk 11]
```diff
@@ -44,6 +45,8 @@ public TransactionEntity map(ResultSet rs, StatementContext ctx) throws SQLExcep
     private ZonedDateTime getZonedDateTime(ResultSet rs, String columnLabel) throws SQLException {
         Timestamp timestamp = rs.getTimestamp(columnLabel);
 
-        return timestamp != null ? ZonedDateTime.ofInstant(timestamp.toInstant(), ZoneOffset.UTC) : null;
+        return Optional.ofNullable(timestamp)
```
Comment (sfount) [inline src/main/java/uk/gov/pay/ledger/transaction/dao/mapper/TransactionMapper.java]: This looks good! Potentially if the Optional was at the level of the method it would benefit by: 
* simplifying the method, it can just return the value 
* having the code calling the method explicitly define how to handle the "doesn't exist" case

If this seems like a useful pattern it could be implemented as part of another PR.

Comment (jankowiakmaria) [reply src/main/java/uk/gov/pay/ledger/transaction/dao/mapper/TransactionMapper.java]: As you've mentioned it can be done as separate PR.

File: src/main/java/uk/gov/pay/ledger/transaction/entity/TransactionEntity.java (hunk 7 above)
[the remark this answers is not in the document]
Comment (jankowiakmaria) [reply src/main/java/uk/gov/pay/ledger/transaction/entity/TransactionEntity.java]: Separate PR: https://github.com/alphagov/pay-ledger/pull/116

File: src/main/java/uk/gov/pay/ledger/transaction/dao/mapper/TransactionMapper.java (hunk 11 above)
Comment (jankowiakmaria) [reply src/main/java/uk/gov/pay/ledger/transaction/dao/mapper/TransactionMapper.java]: https://github.com/alphagov/pay-ledger/pull/120
